### ISSUE TITLE
[RyuJIT] Expression a % 0 sometimes generates incorrect code.

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#8648]
The following example:
```C#
[MethodImpl(MethodImplOptions.NoInlining)]
static int Test(sbyte x)
{
    return x % 0;
}
```
generates
```asm
G_M55886_IG02:
       480FBEC1             movsx    rax, cl
G_M55886_IG03:
       C3                   ret
```
Of course, this doesn't generate a `DivideByZeroException` exception as it should.

This is partially caused by a recent change I did that transforms `x % c` into `x - (x / c) * c` if c is not a power of 2. 0 is not a power of 2 but the transform should not be done when c is 0. ARM64 appears to do the same transform unconditionally so probably it is also affected by this. I think we should simply transform `x % 0` into `x / 0`.

However, the actual bug is somewhere else. The following IL code triggers it as well:
```
.method private hidebysig static int32 
        Test(int8 arg) cil managed noinlining
{
        .maxstack  8
        ldarg.0
        ldarg.0
        ldc.i4.0
        div
        ldc.i4.0
        mul
        sub
        ret
}
```
The C# compiler doesn't usually produce this kind of code, it drops `x * 0`. It's that multiplication that's probably mishandled by the JIT, it tries to drop it and sometimes it drops the division as well. When it works correctly the generated code looks like this:
```asm
G_M55886_IG02:
       480FBEC9             movsx    rcx, cl
       4533C0               xor      r8d, r8d
       8BC1                 mov      eax, ecx
       99                   cdq
       41F7F8               idiv     edx:eax, r8d
       8BC1                 mov      eax, ecx
```

I'll look into this. 
category:correctness
theme:cse
skill-level:intermediate
cost:medium

**Comment by mikedn** [comment]
This appears to be a bug in VN or CSE and the multiplication/division by 0 are just the triggers, there may be others. To reproduce this one needs 
* an expression `E`
* `S`, a subexpression of `E` (without side effects)
* `X`, another subexpression of `E` that has an exception side effect
* `VN(S) = VN(E)`

CSE now replaces `E` with `S` because they have the same VN. `E` has an exception side effect inherited from `X` but CSE ignores exceptions and drops `E` together with `X` and its associated side effect.

Now, I'm not sure what exactly the bug is. Is CSE wrong that it ignores exception side effects? Or is VN wrong to assign the same VN to both `S` and `E` considering that one has side effects and the other does not?

Example:
```C#
class Program
{
    static readonly int z = 0; // prevent the C# compiler from dropping multiply by 0
    int y;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Test(Program p, int x)
    {
        return (x * 2) + (p.y * z);
    }

    static void Main()
    {
        Console.WriteLine(z.ToString());
        Console.WriteLine(Test(null, 5).ToString());
    }
}
```
This program is supposed to crash with `NullReferenceException` but it doesn't because `(p.y * z)` has been eliminated.
Before CSE we have:
```
***** BB01, stmt 1
     ( 10, 10)              [000010] ------------             *  STMT      void  (IL 0x000...0x010)
N011 ( 10, 10)              [000009] ---XG-------             \--*  RETURN    int    $201
N008 (  1,  1)              [000006] ------------                |     /--*  CNS_INT   int    0 $40
N009 (  5,  5)              [000007] ---XG-------                |  /--*  COMMA     int    $40
N007 (  4,  4)              [000005] ---XG-------                |  |  \--*  IND       int    <l:$102, c:$200>
N005 (  1,  1)              [000011] ------------                |  |     |  /--*  CNS_INT   long   8 field offset Fseq[y] $140
N006 (  2,  2)              [000012] -------N----                |  |     \--*  ADD       byref  $180
N004 (  1,  1)              [000004] ------------                |  |        \--*  LCL_VAR   ref    V00 arg0         u:2 (last use) $80
N010 (  9,  9) CSE #01 (use)[000008] ---XG-------                \--*  ADD       int    $100
N002 (  1,  1)              [000002] ------------                   |  /--*  CNS_INT   int    1 $41
N003 (  3,  3) CSE #01 (def)[000003] ------------                   \--*  LSH       int    $100
N001 (  1,  1)              [000001] ------------                      \--*  LCL_VAR   int    V01 arg1         u:2 (last use) $c0
```
and after CSE we have:
```
     (  5,  5) [000010] ------------             *  STMT      void  (IL 0x000...0x010)
N008 (  5,  5) [000009] -A--G-------             \--*  RETURN    int    $201
N006 (  1,  1) [000017] ------------                |  /--*  LCL_VAR   int    V03 cse0          $100
N007 (  4,  4) [000018] -A----------                \--*  COMMA     int    $100
N002 (  1,  1) [000002] ------------                   |     /--*  CNS_INT   int    1 $41
N003 (  3,  3) [000003] ------------                   |  /--*  LSH       int    $100
N001 (  1,  1) [000001] ------------                   |  |  \--*  LCL_VAR   int    V01 arg1         u:2 (last use) $c0
N005 (  3,  3) [000014] -A------R---                   \--*  ASG       int    $VN.Void
N004 (  1,  1) [000013] D------N----                      \--*  LCL_VAR   int    V03 cse0          $100
```

As far as I can tell the VN assigned to the indir doesn't include exceptions so the comma ends up with the same VN as 0. In turn that results in shift and add having the same VN as well. I tried to add `VNF_NullPtrExc` to the indir and then it works fine but it seems like a heavy handed solution, presumably there's another way.


@RussKeldorph Probably someone from the team who's familiar with VN should take a look at this. I have no idea how it is supposed to behave in this case.

**Comment by mikedn** [comment]
@RussKeldorph bug label?

**Comment by sdmaclea** [comment]
@mikedn I believe this is likely responsible for the following two failures in [Arm64/Unix].  Especially since these tests were in @CarolEidt #8642 PR title.
```
JIT/Regression/JitBlue/DevDiv_359736/DevDiv_359736_do/DevDiv_359736_do.sh
JIT/Regression/JitBlue/DevDiv_359736/DevDiv_359736_ro/DevDiv_359736_ro.sh
```

**Comment by mikedn** [comment]
@sdmaclea Yep, it's possible. These tests run fine on x64 but the relevant code is different. On x64 `a % b` is **sometimes** morphed into `a - (a / b) * b` and that can trigger this bug. On ARM64 `a % b` is **always** morphed (because there's no `mod` instruction) so there's higher chance to trigger this bug.

**Comment by pgavlin** [comment]
I can't repro this with the C# listed above, but I can repro what is essentially the same problem with this source:

```
using System.Runtime.CompilerServices;

class Program
{
    int y;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    static int Test(Program p, int x, int z)
    {
        return (x * 2) + (p.y * z);
    }

    static int Main()
    {
        return Test(null, 5, 0);
    }
}
```

In this instance, constprop folds away the tree that ought to produce the null reference:

```
*************** In optAssertionPropMain()
Blocks/Trees at start of phase

-------------------------------------------------------------------------------------------------------------------------------------
BBnum         descAddr ref try hnd preds           weight   [IL range]      [jump]      [EH region]         [flags]
-------------------------------------------------------------------------------------------------------------------------------------
BB01 [00000183433256C0]  1                              1   [000..009)        (return)                     i label target 
-------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [000..009) (return), preds={} succs={}

***** BB01, stmt 1
     (  7,  7) [000011] ------------             *  stmtExpr  void  (IL   ???...  ???)
N007 (  7,  7) [000010] ---XG-------             \--*  return    int    $101
N005 (  1,  1) [000015] ------------                |  /--*  const     int    10 $43
N006 (  6,  6) [000024] ---XG-------                \--*  +         int    $43
N003 (  1,  1) [000022] ------------                   |  /--*  const     int    0 $42
N004 (  4,  4) [000023] ---XG-------                   \--*  comma     int    $42
N002 (  3,  3) [000021] ---XG-------                      \--*  indir     int    <l:$c0, c:$100>
N001 (  1,  1) [000027] ------------                         \--*  const     byref  0 $VN.Null

-------------------------------------------------------------------------------------------------------------------
After constant propagation on [000024]:
     (  7,  7) [000011] ------------             *  stmtExpr  void  (IL   ???...  ???)
N007 (  7,  7) [000010] ---XG-------             \--*  return    int    $101
N006 (  6,  6) [000024] ------------                \--*  const     int    10 $43
optVNAssertionPropCurStmt morphed tree:
N002 (  2,  2) [000010] ----G-------             *  return    int    $101
N001 (  1,  1) [000024] ------------             \--*  const     int    10 $43
```

**Comment by mikedn** [comment]
> I can't repro this with the C# listed above

Odd, I still can reproduce that. The example is compiled with the C# compiler that ships with VS2017 using `/o+`. Here's a full JIT [dump](https://gist.github.com/mikedn/2d609be8fa10e7505a888048d68bf316) of the method in case it helps.

> In this instance, constprop folds away the tree that ought to produce the null reference:

Erm, so in your example it's assertion prop that removes the tree, not CSE? Fun. So I would guess that VN is wrong in the way it assigns VNs to trees with exceptions since this affects more than CSE.

**Comment by pgavlin** [comment]
> Odd, I still can reproduce that. The example is compiled with the C# compiler that ships with VS2017 using /o+. Here's a full JIT dump of the method in case it helps.

Interesting. It looks like the runtime is making different choices w.r.t. how the static field should be accessed:

```
Importing BB01 (PC=000) of 'Program:Test(ref,int):int'
    [ 0]   0 (0x000) ldarg.1
    [ 1]   1 (0x001) ldc.i4.2 2
    [ 2]   2 (0x002) mul
    [ 1]   3 (0x003) ldarg.0
    [ 2]   4 (0x004) ldfld 04000002
    [ 2]   9 (0x009) ldsfld 04000001
    [ 3]  14 (0x00e) mul
    [ 2]  15 (0x00f) add
    [ 1]  16 (0x010) ret

               [000017] ------------             *  stmtExpr  void  (IL 0x000...  ???)
               [000016] --CXG-------             \--*  return    int   
               [000006] ----G-------                |        /--*  field     int    z
               [000013] --CXG-------                |     /--*  comma     int   
               [000012] H-CXG-------                |     |  \--*  call help long   HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000008] ------------ arg0           |     |     +--*  const     long   0x7ffce6d65e20
               [000009] ------------ arg1           |     |     \--*  const     int    1
               [000014] --CXG-------                |  /--*  *         int   
               [000005] ---XG-------                |  |  \--*  field     int    y
               [000004] ------------                |  |     \--*  lclVar    ref    V00 arg0         
               [000015] --CXG-------                \--*  +         int   
               [000002] ------------                   |  /--*  const     int    2
               [000003] ------------                   \--*  *         int   
               [000001] ------------                      \--*  lclVar    int    V01 arg1         

New BlockSet epoch 1, # of blocks (including unused BB00): 2, bitset array size: 1 (short)
```

> Erm, so in your example it's assertion prop that removes the tree, not CSE? Fun. So I would guess that VN is wrong in the way it assigns VNs to trees with exceptions since this affects more than CSE.

It looks like there are two possibilities: either then VN is wrong, or the checks we run before mutating trees are wrong. @briansull has me mostly convinced that it is the former, as in principle VNs should take into account both the value and the exceptions produced by a tree. In these cases that clearly isn't happening (indeed, neither [the code that numbers `(ind arbitrary_addr)`](https://github.com/dotnet/coreclr/blob/master/src/jit/valuenum.cpp#L6789-L6796) nor [the code that numbers `(div x)`](https://github.com/dotnet/coreclr/blob/master/src/jit/valuenum.cpp#L6884-L6895) seems to include the appropriate exceptions in the VNs).

**Comment by pgavlin** [comment]
@JosephTremoulet do you have any additional insight on this?

**Comment by mikedn** [comment]
> Interesting. It looks like the runtime is making different choices w.r.t. how the static field should be accessed:

Maybe you omitted this code from my example:
```C#
Console.WriteLine(z.ToString());
```
It's there to ensure that the static ctor has ran before the JIT compiles `Test`.

**Comment by pgavlin** [comment]
> Maybe you omitted this code from my example:

Yep, that's it--good catch.

**Comment by JosephTremoulet** [comment]
> @JosephTremoulet do you have any additional insight on this?

Not really...

> @briansull has me mostly convinced that it is the former, as in principle VNs should take into account both the value and the exceptions produced by a tree

Yes, I think the reason value numbers include exception info is to avoid this sort of bug.

>the code that numbers (ind arbitrary_addr) [doesn't] include the appropriate exceptions in the VN

When I added that code, I copied its [exception processing](https://github.com/dotnet/coreclr/blob/c52a9f7a3019d408cd03aef9e5fb33263cb047ba/src/jit/valuenum.cpp#L6795) from the other cases just above.  Looking closely now at how `addrXvnp` gets computed, it appears that it's the exceptions from the address expression and hence doesn't include the nullref.  So then probably null ref needs to get added to the exception set for most of the cases in this function?

**Comment by pgavlin** [comment]
> So then probably null ref needs to get added to the exception set for most of the cases in this function?

FWIW, there only appear to be four (!) uses of `VNF_NullPtrExc` in the compiler:
- [valuenum.cpp:1568](https://github.com/dotnet/coreclr/blob/master/src/jit/valuenum.cpp#L1562-L1568): here we are numbering `(ARR_LENGTH null)`, which clearly produces a null pointer exception.
- [valuenum.cpp:6638, valuenum.cpp:6651](https://github.com/dotnet/coreclr/blob/master/src/jit/valuenum.cpp#L6625-L6657): in these cases we are numbering something that was imported as `(IND (ADD array_addr array_len_offset))` and the value number for `(ADD array_addr array_len_offset)` (or whatever it has been transformed into) is a constant. We probably expect that constant to be something derived from `null`, so we add in the null pointer exception.
- [valuenum.cpp:6937](https://github.com/dotnet/coreclr/blob/master/src/jit/valuenum.cpp#L6932-L6939): we are numbering an explicit null check, which clearly may produce a null pointer exception.

It does not appear that we include the null pointer exception in any other case that may produce a null pointer exception (e.g. field accesses). This makes me wonder whether or not the omission of the null pointer exception was deliberate: perhaps adding these exceptions to the relevant value numbers causes us to miss opportunities for optimization? It seems to me that folding these exceptions into value numbers would require that we renumber after proving non-null assertions, which I don't think we currently do.

Thoughts?

**Comment by mikedn** [comment]
> This makes me wonder whether or not the omission of the null pointer exception was deliberate: perhaps adding these exceptions to the relevant value numbers causes us to miss opportunities for optimization?

Perhaps the thinking was that `GTF_EXCEPT` is enough to avoid issues like this. And at least for CSE it usually is, it's just this somewhat peculiar case of a subexpression that ends up replacing its parent and its side effects that causes problems.

**Comment by pgavlin** [comment]
> Perhaps the thinking was that `GTF_EXCEPT` is enough to avoid issues like this

I thought that might be the case at first, but neither assertion prop nor CSE check the `GTF_EXCEPT` flag when performing replacements. Instead, they use `GTF_PERSISTENT_SIDE_EFFECTS`, which only includes calls and assignments. While it's possible that this is an oversight, it certainly seems deliberate. @briansull may be able to confirm this.

**Comment by JosephTremoulet** [comment]
> for CSE it usually is

Can you explain that in a bit more detail?  I went looking for something like that, but found that when `PerformCSE` [extracts side-effects](https://github.com/dotnet/coreclr/blob/0d02d3bc5cab6ea40c4c431cbbcc7e14a5332835/src/jit/optcse.cpp#L2021-L2023), it uses the `GTF_PERSISTENT_SIDE_EFFECTS_IN_CSE` flag, from which `GTF_EXCEPT` is [conspicuously absent](https://github.com/dotnet/coreclr/blob/0d02d3bc5cab6ea40c4c431cbbcc7e14a5332835/src/jit/gentree.h#L757).  And [similarly](https://github.com/dotnet/coreclr/blob/0d02d3bc5cab6ea40c4c431cbbcc7e14a5332835/src/jit/optcse.cpp#L2416) for `optIsCSEcandidate`.

> perhaps adding these exceptions to the relevant value numbers causes us to miss opportunities for optimization?

Notice the unpack/value-number/repack pattern.  The value number would appear to be a pair of (value if no exception happened, sequence of conditions that would cause exceptions).  So we should get a match whenever the possibly-null dereferenced pointers have the same values in each case, but not when they don't (which is what's happening in this bug).  So it may work out ok, or I'm overlooking something?

>would require that we renumber after proving non-null assertions

I'm not following why that would be a requirement.  Do you mean for correctness or for optimal codegen?  So long as we *never* embed "oh I knew this wasn't null" information based on controlling predicates into the value numbers, then the version with the exception annotations should work as a sort of normal form, right?  I suppose we'd miss the cases that are like the ones in this bug (that we should avoid CSEing), when they happen under predicates that have proven the exceptions won't happen (so it turns out CSEing them is sound), but that doesn't seem like much of a problem,

**Comment by pgavlin** [comment]
> So we should get a match whenever the possibly-null dereferenced pointers have the same values in each case, but not when they don't (which is what's happening in this bug). So it may work out ok, or I'm overlooking something?

The problem isn't matching the dereferences themselves, it's matching whatever consumes the value produced by the dereference: because the consumer's value number now contains an exception, I can imagine that we might miss CSEs, constant folds, etc. that we might get if we did not add the exception to the VN but instead checked for `GTF_EXCEPT` when rewriting the IR.

> Do you mean for correctness or for optimal codegen?

I mean for optimal codegen, not for correctness.

**Comment by JosephTremoulet** [comment]
> it's matching whatever consumes the value produced by the dereference: because the consumer's value number now contains an exception, I can imagine that we might miss CSEs, constant folds, etc.

If the best value number (ignoring exceptions) that we have for an `IND (p)` node is `VN_IND (<p's VN>)` (meaning that we don't really know what value was loaded), then any consumers would be have to be functions of loads of things that are value-equivalent to `p`, so I'd expect them all to carry the `exception if <p's VN> is null` annotation.

On the other hand, if we have a better value number for an `IND (p)` node, then we must have seen a store to `p` (or another pointer with p's VN) and be getting the value number from there.  Sounds to me like the code that propagates a value-number from a load to a store should grab the exceptions from the stored value, not the stored/loaded-through pointer.  So then consumers would be seeing something that is a function of the stored value, not something with the `exception if <p's VN> is null` annotation.

**Comment by pgavlin** [comment]
You're right.

I must have been concerned primarily with comma cases. If the `IND` appears on the LHS of a comma node, its value is discarded but its set of exceptions is not: the value number for the comma node is the value number of its RHS + the union of the set of exceptions in the LHS and the RHS. The repro I posted is a motivating example: we may not want to lose the ability to constant fold the add just because one of its operands is an exception-producing comma node.

**Comment by mikedn** [comment]
> Can you explain that in a bit more detail? I went looking for something like that, but found that when PerformCSE extracts side-effects, it uses the GTF_PERSISTENT_SIDE_EFFECTS_IN_CSE flag, from which GTF_EXCEPT is conspicuously absent. And similarly for optIsCSEcandidate.

Yes, you're right. In fact I mentioned this in my second post but I completely forgot, it's been a while. I was initially surprised of CSE's disregard of `GTF_EXCEPT` but then concluded that it simply expects that all expressions produce the same exception, if any. There's no problem doing CSE on
```C#
return (p.y * z) + (p.y * z);
```
because you get something like
```C#
t = (p.y * z);
return t + t;
```
and this then (correctly) reduces to:
```C#
p.y;
return 0;
```

We only get into trouble when one of the expression doesn't throw and has the same value number as the one that does throw. That seems like a rather rare situation, if a VN depends on a load it's difficult to obtain the same VN without doing a load. I tried various things but the only way I have been able to reproduce this was to involve multiplication by 0 in a subexpression (note that the above example also multiplies by 0 but works fine).

I suppose that in the case of CSE this may be fixed by checking that the expression that remains after CSE still has `GTF_EXCEPT` if any of the eliminated expressions have `GTF_EXCEPT`. Though I suspect that some weird enough tree would show us a NullRefException being replaced by a DivByZeroException...

**Comment by JosephTremoulet** [comment]
> I must have been concerned primarily with comma cases. If the IND appears on the LHS of a comma node, its value is discarded but its set of exceptions is not: the value number for the comma node is the value number of its RHS + the union of the set of exceptions in the LHS and the RHS. The repro I posted is a motivating example: we may not want to lose the ability to constant fold the add just because one of its operands is an exception-producing comma node.

That makes sense.  If you're suggesting a course of action, it's not clear to me what the suggestion is.  If you're just thinking through the issues, then I think we're on the same page.

I still think that the current design of the value numbering is to include the VN exception annotations for stuff like this, and that therefore "the bug" is that it's not attaching the nullref annotations to those loads.

If we find motivation here to change the design, fine, but I'm not convinced that we have/will.  Considering the comma case, where we have some node (Expr1, Expr2) where Expr1 has side-effects like nullref/divZero:
 - As far as optimizing the comma itself goes, CSE and constant prop will still rewrite Expr2, which is really what we want (it optimizes the expression while preserving the side-effects)
 - As far as rewriting consumers goes:
    - Supposing the comma feeds into constant-foldable arithmetic:
      - The constant-folding code in value-numbering [unpacks](https://github.com/dotnet/coreclr/blob/c52a9f7a3019d408cd03aef9e5fb33263cb047ba/src/jit/valuenum.cpp#L7029) exceptions from incoming VNs before [attempting to fold](https://github.com/dotnet/coreclr/blob/c52a9f7a3019d408cd03aef9e5fb33263cb047ba/src/jit/valuenum.cpp#L7069) and then [re-attaches](https://github.com/dotnet/coreclr/blob/c52a9f7a3019d408cd03aef9e5fb33263cb047ba/src/jit/valuenum.cpp#L7071) the exception annotations, so we'll still know what folded constants are produced
      - From scanning the code, it looks like [constant prop](https://github.com/dotnet/coreclr/blob/c52a9f7a3019d408cd03aef9e5fb33263cb047ba/src/jit/assertionprop.cpp#L2380-L2384) will bail on the resulting nodes even though it knows what constant they folded to; I'd think that can/should be updated to unpack the exceptions before checking if the VN is constant, and to extract side-effects (including GTF_EXCEPT) when replacing trees with constants.
   - Supposing the comma feeds into larger expressions that should be CSEable (including by way of SSA defs/uses):
     - It does seem like we wouldn't get those CSEs
     - I don't think we want to change CSE to preserve GTF_EXCEPT, since the flip side of not getting those CSEs is that the CSEs we do perform have guarantees that the sub-expressions are safe
     - I wonder if we want expression shaping to reassociate commas root-ward, so that side-effects from the first expr will pollute as little as possible...

**Comment by pgavlin** [comment]
> If you're just thinking through the issues, then I think we're on the same page.

Yes, I've been thinking this through. Good to hear that we're on the same page.

> I still think that the current design of the value numbering is to include the VN exception annotations for stuff like this, and that therefore "the bug" is that it's not attaching the nullref annotations to those loads.
>
> If we find motivation here to change the design, fine, but I'm not convinced that we have/will

At the moment I think I agree. The only thing that is really at all concerning is the comma issue we've been discussing. If this issue turns out to be especially prevalent (and I am not sure whether or not it is), then we may need to rethink our strategy a bit (up to and including your suggestion re: reassociating commas).

For now I will try and find the places in which we ought to be adding exceptions and see how those changes affect our ability to optimize.

**Comment by briansull** [comment]
Here is an additional case to ponder:

After adding in the proper VNF_NullPtrExc for the left-hand-side of a comma
We will get the same value  number for these two trees:

(*p, (x + y))   and (*q, (x + y))

The left-hand-side of the comma's differs as follows:
One dereferences p and discards the result and the other dereferences q and discards the result.

**Comment by JosephTremoulet** [comment]
@briansull , `VNF_NullPtrExc` [takes one source argument](https://github.com/dotnet/coreclr/blob/408390ebf141738118924db3694bc487c8634f56/src/jit/valuenumfuncs.h#L55), the value number of the pointer, I presume for this reason.

**Comment by JosephTremoulet** [comment]
But `VNF_DivideByZeroExc` doesn't seem to take the dividend as a source...

**Comment by BruceForstall** [comment]
@mikedn @briansull Any thoughts on what should be done about this now? It does appear to be the cause of several issues, including #10111.

**Comment by briansull** [comment]
I am testing a fix for this issue.

**Comment by BruceForstall** [comment]
@briansull Does https://github.com/dotnet/coreclr/pull/19845 fix this?

**Comment by briansull** [comment]
No that is prequel to the actual fix

**Comment by briansull** [comment]
Fixed with #20129
